--- libimobiledevice-1.1.1/src/lockdown.c	2011-04-11 23:03:32.000000000 +0800
+++ libimobiledevice-1.1.1/src/lockdown.c	2011-10-17 17:29:42.559172832 +0800
@@ -86,7 +86,9 @@
 
 	plist_t result_node = plist_dict_get_item(dict, "Result");
 	if (!result_node) {
-		return ret;
+        /* iOS 5: the 'Result' key is not present anymore.
+         * Just assume success here */
+		return RESULT_SUCCESS;
 	}
 
 	plist_type result_type = plist_get_node_type(result_node);
