--- linux-2.6.26/drivers/net/wireless/rt2860v2/common/cmm_info.c	2014-04-10 17:07:26.474755765 +0800
+++ linux-2.6.26/drivers/net/wireless/rt2860v2/common/cmm_info.c	2014-04-15 18:21:40.205857627 +0800
@@ -3304,12 +3304,13 @@
 		}
 	}
 
+#ifndef CONFIG_AMIT_CDE570_PROX
 	wrq->u.data.length = sizeof(RT_802_11_MAC_TABLE);
-	if (copy_to_user(wrq->u.data.pointer, pMacTab, wrq->u.data.length))
+    if (copy_to_user(wrq->u.data.pointer, pMacTab, wrq->u.data.length))
 	{
 		DBGPRINT(RT_DEBUG_TRACE, ("%s: copy_to_user() fail\n", __FUNCTION__));
 	}
-
+#endif
 
 /*	msg = kmalloc(sizeof(CHAR)*(MAX_LEN_OF_MAC_TABLE*MAC_LINE_LEN), MEM_ALLOC_FLAG);*/
 	os_alloc_mem(NULL, (UCHAR **)&msg, sizeof(CHAR)*(MAX_LEN_OF_MAC_TABLE*MAC_LINE_LEN)+100);
@@ -3345,6 +3345,13 @@
 
 		}
 	} 
+#ifdef CONFIG_AMIT_CDE570_PROX
+    wrq->u.data.length = strlen(msg);
+    if (copy_to_user(wrq->u.data.pointer, msg, wrq->u.data.length))
+	{
+		DBGPRINT(RT_DEBUG_TRACE, ("%s: copy_to_user() fail\n", __FUNCTION__));
+	}
+#endif
 	/* for compatible with old API just do the printk to console*/
 
 	DBGPRINT(RT_DEBUG_TRACE, ("%s", msg));
